Move active EP allocation to common code

shefty opened the pull request:
Replace individual implementations of alloc_ep_res() with
a shared implementation for allocating an active EP and
associated resources.

This patch also fixes incorrect usage of determining the
'max_credits' of requests that can be posted to the transmit
queue of the EP.

Private usages of 'transfer_size' are also replaced with
the common transfer size option.

Signed-off-by: Sean Hefty <sean.hefty@intel.com>

bturrubiates commented:
There's still an issue with the `usNIC` provider that causes `rdm_pingpong` and `msg_pingpong` to fail. The sockets provider is also failing a couple of tests after this.

shefty commented:
I didn't see a failure with the sockets provider in my tests.  Personally, the more I work on cleaning things up, the tests are wrong in several areas.  The msg prefix is incorrect.  There were hard-coded assumptions about the size of the available queues.  The memory registration was wrong.  The tests won't work between 32 and 64 bit systems, or different endianness, etc. Bah.

Can you report the test parameters that you used?  Note that I had to update to the latest version of libfabric to not have the sockets provider crash in my testing.

bturrubiates commented:
I get a hang on `fi_rdm_tagged_pingpong` with the `sockets` provider. Invocation was simly `./fi_rdm_tagged_pingpong -f sockets -s {server}` and on client `./fi_rdm_tagged_pingpong -f sockets {server}`. 

The `usNIC` problems are likely not related to the test. I think your test just uncovered an issue in the `usNIC` provider. I'm glad the tests are getting looked at, I agree that they could be in better shape.

shefty commented:
If I don't use the -I option, I can see it crash.  Looking at the test, it's wrong when FI_CONTEXT is set.  It reuses the same content for all sends without waiting for a completion.  I'll remove it's support for FI_CONTEXT until the code can be restructured to work.  I'll check other tests for correctness as well.

I'm not sure if this is the issue or not.  Even running inside the debugger, things are so hosed, that I don't get a meaningful backtrace.  It looks like the instruction pointer was set to NULL.  :/

shefty commented:
Fixed crash in fi_rdm_tagged_pingpong.  It was a stack corruption as a result of changing the CQ format, but not updating the data structure.  Also added a patch to remove FI_CONTEXT support from 2 tests that do not correctly support it.

bturrubiates commented:
That fixed the issues I was seeing with the `sockets` provider. :+1: 
The `usNIC` provider is still having issues, but I'll investigate those.

bturrubiates commented:
Oh, found the problem. I was looking in the wrong spot for a while. Here's a patch I can apply to make this pass for the `usNIC` provider: 

```
diff --git a/common/shared.c b/common/shared.c
index 90c359a..17af359 100644
--- a/common/shared.c
+++ b/common/shared.c
@@ -155,7 +155,7 @@ int ft_alloc_active_res(struct fi_info *fi)
        }

        if (opts.options & FT_OPT_TX_CQ) {
-               cq_attr.size = fi->tx_attr->size;
+               cq_attr.size = (fi->tx_attr->size << 1);
                ret = fi_cq_open(domain, &cq_attr, &txcq, &txcq);
                if (ret) {
                        FT_PRINTERR("fi_cq_open", ret);
@@ -172,7 +172,7 @@ int ft_alloc_active_res(struct fi_info *fi)
        }

        if (opts.options & FT_OPT_RX_CQ) {
-               cq_attr.size = fi->rx_attr->size;
+               cq_attr.size = (fi->rx_attr->size << 1);
                ret = fi_cq_open(domain, &cq_attr, &rxcq, &rxcq);
                if (ret) {
                        FT_PRINTERR("fi_cq_open", ret);
```

As for the actual issue: the old code had a much larger CQ than the number of transmit credits. The test only reads completions when it is out of transmit credits. In this PR the number of transmit credits and the number of CQ entries are the same. This means that when the transmit credits run out and it's time to read a CQ event the `usNIC` providers CQ implemented as a circular buffer wraps around to the beginning and then returns `-FI_EAGAIN` due to `tail == beginning_of_queue`. I'll be working on that problem separately, but is this test intentionally changing the semantics (`cq_size` == `fi_tx_size` rather than `cq_size` == `fi_tx_size << 1`)?

shefty commented:
>         if (opts.options & FT_OPT_TX_CQ) {
> -               cq_attr.size = fi->tx_attr->size;
> +               cq_attr.size = (fi->tx_attr->size << 1);

I noticed that the old code did this, and this looks wrong to me.  If anything, we should adjust by 1, not double the size of the CQ.  Though I don't think we should even need to increment by 1.  If the send queue has 100 entries, we only need 100 entries on the CQ.

So, yes, this change was intentional.  I don't mind bumping this to work around a provider bug, but we should document that that is why the change is made, so that it can be reverted at a later date.  Specifically because it can find those bugs.

bturrubiates commented:
Sure, I'll work on fixing the provider bug.

shefty commented:
Do you need a follow on patch to support usnic, or are you okay fixing the provider?

bturrubiates commented:
I'll fix the provider
